openedx/frontend-app-course-authoring
# [FC-0036] feat: Add content tags tree state + editing

## Description

This PR adds the editing functionality for content tags. You are able to select/deselect tags from the dropdowns in each taxonomy and the tags tree structure is updated, requests are sent to the backend, everything gets automatically updated and the state persists.

When a tag is select, all its ancestors automatically become 'implicit' and are unselected and cannot be selected. If an explicit tag is unselected and the parent (or ancestor) no longer has explicit children, all ancestor tags will also be removed.

![Screen Shot 2023-11-21 at 7 26 53 PM](https://github.com/openedx/frontend-app-course-authoring/assets/6829768/35e265f2-3cd9-4e55-b0c5-046f29cb163f)

## Supporting information

Related Tickets:
- https://github.com/openedx/modular-learning/issues/122

## Testing instructions

1. Run you local dev stack.
2. Next, navigate to http://localhost:18010/admin/waffle/flag/ and create a new flag: new_studio_mfe.use_tagging_taxonomy_list_page make sure it is active for everyone
3. If you don't already have sample taxonomy/tags data, follow the instructions in this repo to generate sample data: https://github.com/open-craft/taxonomy-sample-data
4. Navigate to the sample courses and open the tags drawer for the units
5. Check that all the taxonomies are shown in the drawer and the applied tags are show
6. Expand some taxonomies and check that the applied tags shows, implicit and explicit
7. Click on "Add Tags" button and confirm that the applied tags are reflected in the dropdown selector as well
8. Confirm that the implicit tags are not clickable
9. Confirm that when you check/uncheck a tag it is reflected in the tags tree
10. Confirm that if you check a tag, all its ancestors become implicit
11. Confirm when you uncheck a tag and it is the only child, all its ancestors get unchecked
12. Confirm when you uncheck a tag and its has siblings, the ancestors do not get removed
13. Confirm that "Load More" button functions correctly and loads more tags
14. Confirm that the tags number gets updated correctly whenever you add/remove tags
15. Confirm that you can remove tags by clicking on the "x" for explicit tags (same behaviors described above should be observed)
---
Private-ref: [FAL-3530](https://tasks.opencraft.com/browse/FAL-3530)
